src/firmware_new/boards/frankie_v1/udp_multicast.c

```diff
@@ -0,0 +1,117 @@
+    unsigned multicast_port;
+    unsigned send_port;
+} typedef multicast_config_t;
+
+/*
+ * Thread that creates a send and recv socket, joins the specified
+ * multicast group and start listening for packets. Increments a counter
+ * on every multicast packet and sends a unicast packet back to the sender.
+ *
+ */
+static void udp_multicast_thread(void *arg)
+{
+    multicast_config_t *config = (multicast_config_t *)arg;
+    struct netbuf *rx_buf, *tx_buf;
+    int msg_count = 0;
+
+    // TODO proper err handling, for now all errs are ignored
+    err_t err;
+
+    // create two new UDP connections on the heap
+    // NOTE: we need two seperate UDP sockets as we will be
+    // receiving multicast but sending unicast packets to avoid
+    // network congestion (peer robots don't need to know the status
+    // of other robots)
+    struct netconn *recvconn = netconn_new(NETCONN_UDP_IPV6);
+    struct netconn *sendconn = netconn_new(NETCONN_UDP_IPV6);
+
+    // bind the socket to the multicast address and port
+    // we then use that connection profile to join the specified
+    // multicast group. For the send socket, we use IP6_ADDR_ANY
+    // which will default to the ETH interface.
+    netconn_bind(recvconn, &config->multicast_address, config->multicast_port);
+    netconn_bind(sendconn, IP6_ADDR_ANY, config->send_port);
+    netconn_join_leave_group(recvconn, &config->multicast_address, NULL, NETCONN_JOIN);
+
+    while (1)
+    {
+        err = netconn_recv(recvconn, &rx_buf);
+
+        if (err == ERR_OK)
+        {
+            tx_buf = netbuf_new();
+            netbuf_alloc(tx_buf, robot_ack_size);
+
+            // Create a stream that reads from the buffer
+            pb_istream_t in_stream =
+                pb_istream_from_buffer((uint8_t *)rx_buf->p->payload, rx_buf->p->tot_len);
+
+            if (pb_decode(&in_stream, control_msg_fields, &control))
+            {
+                // update proto
+                ack.msg_count = msg_count++;
+
+                // serialize proto
+                pb_ostream_t stream = pb_ostream_from_buffer(buffer, sizeof(buffer));
+                pb_encode(&stream, robot_ack_fields, &ack);
+
+                // package payload and send over udp
+                tx_buf->p->payload = buffer;
+                netconn_sendto(sendconn, tx_buf, (const ip_addr_t *)&(rx_buf->addr),
+                               config->send_port);
+            }
+            netbuf_delete(tx_buf);
+        }
+        netbuf_delete(rx_buf);
+    }
+}
+
+void udp_multicast_init(const char *multicast_address, unsigned multicast_port,
+                        unsigned send_port)
+{
+    multicast_config_t *config = malloc(sizeof(multicast_config_t));
+
+    *config = (multicast_config_t){
+        .multicast_port = multicast_port,
+        .send_port      = send_port,
+    };
+
+    // create multicast address and port from arguments
+    // htons swaps bytes to correct the endianness of the port,
```

Comment:
what is "htons"? typo or real terminology :thinking:

Reply:
ah that was there before when we weren't using net conn, relic of the past, removed